p4-16/psa/PSA.mdk

```diff
@@ -950,41 +941,35 @@ used in the second pass. See Section [#sec-open-metadata] for the
 discussion.
 
 A PSA implementation provides a configuration bit `resubmit` to the
-PRE to enable the resubmission mechanism. If `resubmit` is set to
-true, the resubmission mechanism resends the original packet with the
-optional attached metadata. If `resubmit` is set to false, the
-resubmission mechanism is disabled and the original packet is not
-resubmitted even if the `add_metadata` method is invoked in the
-deparser.
-
-```
-[INCLUDE=psa.p4:Resubmit_extern]
-
-```
+PRE to enable the resubmission mechanism. If true, the original packet
+is resubmitted with the optional resubmit metadata. If false, the
+resubmission mechanism is disabled and no assignments to
+`resubmit_meta` should be performed.
 
 ## Packet Recirculation {#sec-recirculate}
 
-Packet recirculation sends the deparsed packet at the end of egress
-pipeline to the ingress parser. Unlike the previous two mechanisms,
-recirculation does not replicate packets. The modified packet itself
-is sent back to the ingress pipeline for further processing.
-
-In PSA, recirculation is controlled by the `recirculate` bit in the
-egress pipeline. If the `recirculate` bit is set to true, the modified
-packet at the end of the pipeline is sent to a special recirculation
-port. The recirculation port overrides the value in `egress_port`.
-
-```
-[INCLUDE=psa.p4:Recirculate_extern]
-
-```
-
-Users can leverage the existing `emit` methods in `packet_out` to
-attach additional metadata to the recirculated packet. Upon receiving
-the recirculated packet in ingress parser, the `packet_path` of the
-recirculated packet is set to `RECIRCULATE`. It is the user's
-responsibility to parse the recirculated packet correctly.
-
+Packet recirculation is a mechanism to repeat both ingress and egress
+processing on a packet.
+
+Packet recirculation happens at the end of the egress pipeline. When a
+packet is recirculated, the packet finishes both ingress and egress
+processing and re-enters the ingress parser after being deparsed by
+the egress deparser. The `ingress_port` of the recirculated packet is
+set to `PORT_RECIRCULATE`. The `packet_path` of the recirculated
+packet is changed to `RECIRCULATE`.
+
+Similar to packet resubmission, packet recirculation also supports
+attaching optional metadata with the recirculated packet. The metadata
+is generated during the first pass through the ingress and egress
+pipeline, and used in the second pass. See Section
+[#sec-open-metadata] for the discussion.
+
+The `recirculate` flag specifies whether a packet should be
+recirculated. If true, the processed packet should be recirculated at
+the end of the egress pipeline with the optional metadata. If false,
+the recirculation mechanism is disabled and no assignments to
+`recirc_meta` should be performed. Some target may restrict the
+`recirculate` flag to be set in the ingress pipeline.
```

Comment:
The `recirculate` flag is not visible or accessible in the ingress pipeline in the current psa.p4 definition.  How do you propose a target require the programmer to assign a value to `recirculate` in the ingress pipeline?

Reply:
Follow-up comment: If someone wants to control whether a recirculate happens in their ingress code, a way to do that is to store in some user-defined metadata field that is carried from ingress to ingress with a normal packet (if the normal packet path is the relevant one here), and then in egress code use that metadata field value to control whether it assigns the value true to the `recirculate` field.